modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m

```diff
@@ -965,73 +966,102 @@ - (NSDragOperation)draggingSourceOperationMaskForLocal:(BOOL)isLocal
 }
 
 // called from Java layer drag handler, triggered by DnD Pasteboard flush
-- (void)startDrag:(NSDragOperation)operation
+- (void)startDrag:(NSDragOperation)operation withItems:(NSArray<NSDraggingItem*>*)items
 {
     DNDLOG("startDrag");
-    self->dragOperation = operation;
+
+    // Set up frames for dragging items
+    for (NSDraggingItem* item in items)
     {
-        NSPoint dragPoint = [self->nsView convertPoint:[self->lastEvent locationInWindow] fromView:nil];
-        NSPasteboard *pasteboard = [NSPasteboard pasteboardWithName:NSDragPboard];
+        NSPoint dragPoint = [self->nsView convertPoint:[self->lastEvent locationInWindow] fromView:nil];//[self->lastEvent locationInWindow];
         NSImage *image = nil;
+        NSRect rect = [item draggingFrame];
+
+        NSPasteboardItem *pbItem = (NSPasteboardItem*)[item item];
+        NSArray<NSPasteboardType> *pbItemTypes = [pbItem types];
 
-        if ([[pasteboard types] containsObject:DRAG_IMAGE_MIME]) {
+        if ([pbItemTypes containsObject:DRAG_IMAGE_MIME]) {
             //Try to init with drag image specified by the user
-            image = [[NSImage alloc] initWithData:[pasteboard dataForType:DRAG_IMAGE_MIME]];
+            image = [[NSImage alloc] initWithData:[pbItem dataForType:DRAG_IMAGE_MIME]];
         }
 
-        if (image == nil && [NSImage canInitWithPasteboard:pasteboard] == YES)
+        if (image == nil && [pbItemTypes containsObject:NSPasteboardTypeFileURL])
```

Comment:
Good catch (which reminds me that I need to get back to your warnings PR). This might misbehave on macOS 10.12 then, but I haven't seen a machine that old in a few years, so it would be difficult to test.

We really should bump the minimum at some point (to at _least_ 10.14, but ideally to 11 to match what Apple currently supports, and to match what the minimum has always been for ARM-based Mac systems), so this seems like the right time to broach that subject.

Reply:
I just sent a [proposal](https://mail.openjdk.org/pipermail/openjfx-dev/2023-May/040392.html) to openjfx-dev to bump the minimum version of macOS to either 10.14 or 11.

Reply:
I tried to cover as many cases as possible with this part since `canInitWithPasteboard` is no longer possible (new API requires not having any items in the Pasteboard, as it puts them in there for you). I can ultimately remove this part, I think it won't hurt the change as it is - only some drag images based from URLs might instead be shown as a generic document icon, which is not a big deal.